[Query service] EntityExistsException when executing a process with cycles

When executing a process that contains a cycle, an exception occurs in query service the second time the task is reached:

![image](https://user-images.githubusercontent.com/26007058/160354570-ac89a6e0-5dc4-4092-bdf7-160f189257a9.png)

```
2022-03-22 09:32:01.792 ERROR [query,5028b105dce87ae9,c52290d99727fccb] 7 --- [eEvents.query-1] o.s.integration.handler.LoggingHandler   : org.springframework.messaging.MessagingException: Exception thrown while invoking QueryConsumerChannelHandler#receive[1 args]; nested exception is javax.persistence.EntityExistsException: A different object with the same identifier value was already associated with the session : [org.activiti.cloud.services.query.model.IntegrationContextEntity#64c3b6d9-a9c2-11ec-bf52-aa11a2095e9d:ServiceTask_1qot7s2:64c3b6db-a9c2-11ec-bf52-aa11a2095e9d], failedMessage=GenericMessage [payload=byte[87735], headers={serviceType=runtime-bundle, amqp_receivedDeliveryMode=PERSISTENT, amqp_receivedExchange=engineEvents, amqp_deliveryTag=21, rootProcessDefinitionKey=Process_skZEn0kcz, amqp_redelivered=false, serviceVersion=, b3=5028b105dce87ae9-d745cfbc897d5d09-1, messagePayloadType=[Lorg.activiti.cloud.api.model.shared.events.CloudRuntimeEvent;, rootProcessDefinitionVersion=1, deploymentId=b617669a-a9be-11ec-bf52-aa11a2095e9d, rootProcessInstanceId=64c3b6d9-a9c2-11ec-bf52-aa11a2095e9d, deploymentName=SpringAutoDeployment, id=23c730c8-a6c6-9664-3c16-7a14228bab8b, sourceData=(Body:'[B@1deab18c(byte[87735])' MessageProperties [headers={serviceType=runtime-bundle, serviceFullName=rb, appName=test-m17, deploymentVersion=1, rootProcessDefinitionKey=Process_skZEn0kcz, serviceName=rb, serviceVersion=, rootProcessDefinitionId=Process_skZEn0kcz:1:b6223c0d-a9be-11ec-bf52-aa11a2095e9d, b3=5028b105dce87ae9-2b32372af29a214c-1, nativeHeaders={}, messagePayloadType=[Lorg.activiti.cloud.api.model.shared.events.CloudRuntimeEvent;, rootProcessDefinitionVersion=1, deploymentId=b617669a-a9be-11ec-bf52-aa11a2095e9d, rootProcessDefinitionName=tomegkoz-tam-public-transp, rootProcessInstanceId=64c3b6d9-a9c2-11ec-bf52-aa11a2095e9d, deploymentName=SpringAutoDeployment, routingKey=engineEvents.rb.test-m17, rootProcessName=tomegkoz-tam-public-transp - Mar 22, 2022, 9:28:11 AM}, timestamp=Tue Mar 22
09:31:58 UTC 2022, messageId=81dadf1c-22d6-b988-6bcf-d38c07706abe, contentType=application/json, contentLength=0, receivedDeliveryMode=PERSISTENT, priority=0, redelivered=false, receivedExchange=engineEvents, receivedRoutingKey=engineEvents, deliveryTag=21, consumerTag=amq.ctag-DdgH84Dny9mI13CLJ4ENGw, consumerQueue=engineEvents.query]), contentType=application/json, timestamp=1647941521743, serviceFullName=rb, appName=test-m17, deploymentVersion=1, deliveryAttempt=3, amqp_consumerQueue=engineEvents.query, serviceName=rb, amqp_receivedRoutingKey=engineEvents, rootProcessDefinitionId=Process_skZEn0kcz:1:b6223c0d-a9be-11ec-bf52-aa11a2095e9d, nativeHeaders={}, amqp_timestamp=Tue Mar 22 09:31:58 UTC 2022, amqp_messageId=81dadf1c-22d6-b988-6bcf-d38c07706abe, rootProcessDefinitionName=tomegkoz-tam-public-transp, amqp_consumerTag=amq.ctag-DdgH84Dny9mI13CLJ4ENGw, routingKey=engineEvents.rb.test-m17, rootProcessName=tomegkoz-tam-public-transp - Mar 22, 2022, 9:28:11 AM}]
	at org.springframework.cloud.stream.binding.StreamListenerMessageHandler.handleRequestMessage(StreamListenerMessageHandler.java:64)
	at org.springframework.integration.handler.AbstractReplyProducingMessageHandler.handleMessageInternal(AbstractReplyProducingMessageHandler.java:136)
	at org.springframework.integration.handler.AbstractMessageHandler.handleMessage(AbstractMessageHandler.java:56)
	at org.springframework.integration.dispatcher.AbstractDispatcher.tryOptimizedDispatch(AbstractDispatcher.java:115)
	at org.springframework.integration.dispatcher.UnicastingDispatcher.doDispatch(UnicastingDispatcher.java:133)
	at org.springframework.integration.dispatcher.UnicastingDispatcher.dispatch(UnicastingDispatcher.java:106)
	at org.springframework.integration.channel.AbstractSubscribableChannel.doSend(AbstractSubscribableChannel.java:72)
	at org.springframework.integration.channel.AbstractMessageChannel.send(AbstractMessageChannel.java:317)
	at org.springframework.integration.channel.AbstractMessageChannel.send(AbstractMessageChannel.java:272)
	at org.springframework.messaging.core.GenericMessagingTemplate.doSend(GenericMessagingTemplate.java:187)
	at org.springframework.messaging.core.GenericMessagingTemplate.doSend(GenericMessagingTemplate.java:166)
	at org.springframework.messaging.core.GenericMessagingTemplate.doSend(GenericMessagingTemplate.java:47)
	at org.springframework.messaging.core.AbstractMessageSendingTemplate.send(AbstractMessageSendingTemplate.java:109)
	at org.springframework.integration.endpoint.MessageProducerSupport.sendMessage(MessageProducerSupport.java:208)
	at org.springframework.integration.amqp.inbound.AmqpInboundChannelAdapter.access$1300(AmqpInboundChannelAdapter.java:69)
	at org.springframework.integration.amqp.inbound.AmqpInboundChannelAdapter$Listener.lambda$onMessage$0(AmqpInboundChannelAdapter.java:374)
	at org.springframework.retry.support.RetryTemplate.doExecute(RetryTemplate.java:329)
	at org.springframework.retry.support.RetryTemplate.execute(RetryTemplate.java:225)
	at org.springframework.integration.amqp.inbound.AmqpInboundChannelAdapter$Listener.onMessage(AmqpInboundChannelAdapter.java:370)
	at org.springframework.amqp.rabbit.listener.AbstractMessageListenerContainer.doInvokeListener(AbstractMessageListenerContainer.java:1665)
	at org.springframework.amqp.rabbit.listener.AbstractMessageListenerContainer.actualInvokeListener(AbstractMessageListenerContainer.java:1584)
	at org.springframework.amqp.rabbit.listener.AbstractMessageListenerContainer.invokeListener(AbstractMessageListenerContainer.java:1572)
	at org.springframework.amqp.rabbit.listener.AbstractMessageListenerContainer.doExecuteListener(AbstractMessageListenerContainer.java:1563)
	at org.springframework.amqp.rabbit.listener.AbstractMessageListenerContainer.executeListener(AbstractMessageListenerContainer.java:1507)
	at org.springframework.amqp.rabbit.listener.SimpleMessageListenerContainer.doReceiveAndExecute(SimpleMessageListenerContainer.java:967)
	at org.springframework.amqp.rabbit.listener.SimpleMessageListenerContainer.receiveAndExecute(SimpleMessageListenerContainer.java:914)
	at org.springframework.amqp.rabbit.listener.SimpleMessageListenerContainer.access$1600(SimpleMessageListenerContainer.java:83)
	at org.springframework.amqp.rabbit.listener.SimpleMessageListenerContainer$AsyncMessageProcessingConsumer.mainLoop(SimpleMessageListenerContainer.java:1291)
	at org.springframework.amqp.rabbit.listener.SimpleMessageListenerContainer$AsyncMessageProcessingConsumer.run(SimpleMessageListenerContainer.java:1197)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: javax.persistence.EntityExistsException: A different object with the same identifier value was already associated with the session : [org.activiti.cloud.services.query.model.IntegrationContextEntity#64c3b6d9-a9c2-11ec-bf52-aa11a2095e9d:ServiceTask_1qot7s2:64c3b6db-a9c2-11ec-bf52-aa11a2095e9d]
	at org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:123)
	at org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:181)
	at org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:188)
	at org.hibernate.internal.SessionImpl.firePersist(SessionImpl.java:766)
	at org.hibernate.internal.SessionImpl.persist(SessionImpl.java:746)
	at jdk.internal.reflect.GeneratedMethodAccessor165.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.springframework.orm.jpa.SharedEntityManagerCreator$SharedEntityManagerInvocationHandler.invoke(SharedEntityManagerCreator.java:311)
	at com.sun.proxy.$Proxy270.persist(Unknown Source)
	at org.activiti.cloud.services.query.events.handlers.IntegrationRequestedEventHandler.handle(IntegrationRequestedEventHandler.java:69)
	at org.activiti.cloud.services.query.events.handlers.QueryEventHandlerContext.lambda$handle$0(QueryEventHandlerContext.java:47)
	at java.base/java.util.Spliterators$ArraySpliterator.forEachRemaining(Spliterators.java:948)
	at java.base/java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:658)
	at org.activiti.cloud.services.query.events.handlers.QueryEventHandlerContext.handle(QueryEventHandlerContext.java:42)
	at org.activiti.cloud.services.query.app.QueryConsumerChannelHandler.receive(QueryConsumerChannelHandler.java:41)
	at org.activiti.cloud.services.query.app.QueryConsumerChannelHandler$$FastClassBySpringCGLIB$$80c8bb00.invoke(<generated>)
	at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:218)
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.invokeJoinpoint(CglibAopProxy.java:783)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:163)
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:753)
	at org.springframework.transaction.interceptor.TransactionInterceptor$1.proceedWithInvocation(TransactionInterceptor.java:123)
	at org.springframework.transaction.interceptor.TransactionAspectSupport.invokeWithinTransaction(TransactionAspectSupport.java:388)
	at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:119)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:753)
	at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:698)
	at org.activiti.cloud.services.query.app.QueryConsumerChannelHandler$$EnhancerBySpringCGLIB$$889b7dee.receive(<generated>)
	at jdk.internal.reflect.GeneratedMethodAccessor411.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.springframework.messaging.handler.invocation.InvocableHandlerMethod.doInvoke(InvocableHandlerMethod.java:169)
	at org.springframework.messaging.handler.invocation.InvocableHandlerMethod.invoke(InvocableHandlerMethod.java:119)
	at org.springframework.cloud.stream.binding.StreamListenerMessageHandler.handleRequestMessage(StreamListenerMessageHandler.java:55)
	... 29 more
```

--- a/activiti-cloud-audit-service/activiti-cloud-starter-audit/src/test/resources/application.properties
+++ b/activiti-cloud-audit-service/activiti-cloud-starter-audit/src/test/resources/application.properties
@@ -15,4 +15,4 @@ authorizations.security-constraints[0].securityCollections[0].patterns[0]=/v1/*
 authorizations.security-constraints[1].authRoles[0]=ACTIVITI_ADMIN
 authorizations.security-constraints[1].securityCollections[0].patterns[0]=/admin/*
 
-spring.jpa.properties.hibernate.generate_statistics=true
+spring.jpa.properties.hibernate.generate_statistics=false

--- a/activiti-cloud-query-service/activiti-cloud-services-query/activiti-cloud-services-query-rest/src/main/java/org/activiti/cloud/services/query/events/handlers/IntegrationRequestedEventHandler.java
+++ b/activiti-cloud-query-service/activiti-cloud-services-query/activiti-cloud-services-query-rest/src/main/java/org/activiti/cloud/services/query/events/handlers/IntegrationRequestedEventHandler.java
@@ -42,13 +42,18 @@ public class IntegrationRequestedEventHandler extends BaseIntegrationEventHandle
         IntegrationContext integrationContext = integrationEvent.getEntity();
         String entityId = IntegrationContextEntity.IdBuilderHelper.from(integrationContext);
 
-        IntegrationContextEntity entity = new IntegrationContextEntity(event.getServiceName(),
-                                                                       event.getServiceFullName(),
-                                                                       event.getServiceVersion(),
-                                                                       event.getAppName(),
-                                                                       event.getAppVersion());
-        entity.setId(entityId);
-        entity.setClientId(integrationContext.getClientId());
+
+        // Activity can be cyclical, so try to find existing before creating a new one
+        IntegrationContextEntity entity = entityManager.find(IntegrationContextEntity.class, entityId);
+        if (entity == null) {
+            entity = new IntegrationContextEntity(event.getServiceName(),
+                event.getServiceFullName(),
+                event.getServiceVersion(),
+                event.getAppName(),
+                event.getAppVersion());
+            entity.setId(entityId);
+            entity.setClientId(integrationContext.getClientId());
+        }
         entity.setClientName(integrationContext.getClientName());
         entity.setClientType(integrationContext.getClientType());
         entity.setConnectorType(integrationContext.getConnectorType());

--- a/activiti-cloud-query-service/activiti-cloud-starter-query/src/test/java/org/activiti/cloud/starter/tests/QueryAdminProcessServiceTasksIT.java
+++ b/activiti-cloud-query-service/activiti-cloud-starter-query/src/test/java/org/activiti/cloud/starter/tests/QueryAdminProcessServiceTasksIT.java
@@ -15,20 +15,53 @@
  */
 package org.activiti.cloud.starter.tests;
 
-import org.activiti.api.runtime.model.impl.*;
+import static org.activiti.cloud.services.query.model.IntegrationContextEntity.ERROR_MESSAGE_LENGTH;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.tuple;
+import static org.awaitility.Awaitility.await;
+
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.util.Arrays;
+import java.util.UUID;
+import org.activiti.api.process.model.BPMNActivity;
+import org.activiti.api.process.model.IntegrationContext;
+import org.activiti.api.process.model.ProcessInstance;
+import org.activiti.api.runtime.model.impl.BPMNActivityImpl;
+import org.activiti.api.runtime.model.impl.BPMNSequenceFlowImpl;
+import org.activiti.api.runtime.model.impl.IntegrationContextImpl;
+import org.activiti.api.runtime.model.impl.ProcessDefinitionImpl;
+import org.activiti.api.runtime.model.impl.ProcessInstanceImpl;
 import org.activiti.cloud.api.process.model.CloudBPMNActivity;
+import org.activiti.cloud.api.process.model.CloudBPMNActivity.BPMNActivityStatus;
 import org.activiti.cloud.api.process.model.CloudBpmnError;
 import org.activiti.cloud.api.process.model.CloudIntegrationContext;
 import org.activiti.cloud.api.process.model.CloudIntegrationContext.IntegrationContextStatus;
 import org.activiti.cloud.api.process.model.CloudServiceTask;
-import org.activiti.cloud.api.process.model.impl.events.*;
-import org.activiti.cloud.services.query.app.repository.*;
+import org.activiti.cloud.api.process.model.impl.events.CloudBPMNActivityCompletedEventImpl;
+import org.activiti.cloud.api.process.model.impl.events.CloudBPMNActivityStartedEventImpl;
+import org.activiti.cloud.api.process.model.impl.events.CloudIntegrationErrorReceivedEventImpl;
+import org.activiti.cloud.api.process.model.impl.events.CloudIntegrationRequestedEventImpl;
+import org.activiti.cloud.api.process.model.impl.events.CloudIntegrationResultReceivedEventImpl;
+import org.activiti.cloud.api.process.model.impl.events.CloudProcessCreatedEventImpl;
+import org.activiti.cloud.api.process.model.impl.events.CloudProcessDeployedEventImpl;
+import org.activiti.cloud.api.process.model.impl.events.CloudProcessStartedEventImpl;
+import org.activiti.cloud.api.process.model.impl.events.CloudSequenceFlowTakenEventImpl;
+import org.activiti.cloud.services.query.app.repository.BPMNActivityRepository;
+import org.activiti.cloud.services.query.app.repository.BPMNSequenceFlowRepository;
+import org.activiti.cloud.services.query.app.repository.IntegrationContextRepository;
+import org.activiti.cloud.services.query.app.repository.ProcessDefinitionRepository;
+import org.activiti.cloud.services.query.app.repository.ProcessInstanceRepository;
+import org.activiti.cloud.services.query.app.repository.ProcessModelRepository;
 import org.activiti.cloud.services.query.model.StringUtils;
 import org.activiti.cloud.services.test.containers.KeycloakContainerApplicationInitializer;
 import org.activiti.cloud.services.test.containers.RabbitMQContainerApplicationInitializer;
 import org.activiti.cloud.services.test.identity.IdentityTokenProducer;
 import org.activiti.cloud.starters.test.EventsAggregator;
 import org.activiti.cloud.starters.test.MyProducer;
+import org.activiti.cloud.starters.test.builder.ProcessInstanceEventContainedBuilder;
 import org.junit.jupiter.api.AfterEach;
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
@@ -44,22 +77,11 @@ import org.springframework.test.annotation.DirtiesContext;
 import org.springframework.test.context.ContextConfiguration;
 import org.springframework.test.context.TestPropertySource;
 
-import java.io.IOException;
-import java.nio.charset.StandardCharsets;
-import java.nio.file.Files;
-import java.nio.file.Paths;
-import java.util.Arrays;
-import java.util.UUID;
-
-import static org.activiti.cloud.services.query.model.IntegrationContextEntity.ERROR_MESSAGE_LENGTH;
-import static org.assertj.core.api.Assertions.assertThat;
-import static org.assertj.core.api.Assertions.tuple;
-import static org.awaitility.Awaitility.await;
-
 @SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
 @TestPropertySource("classpath:application-test-admin.properties")
 @DirtiesContext
-@ContextConfiguration(initializers = { RabbitMQContainerApplicationInitializer.class, KeycloakContainerApplicationInitializer.class})
+@ContextConfiguration(initializers = {RabbitMQContainerApplicationInitializer.class,
+    KeycloakContainerApplicationInitializer.class})
 public class QueryAdminProcessServiceTasksIT {
 
     private static final String ERROR_MESSAGE = "An error occurred consuming ACS API with inputs {targetFolder={}, action=CREATE_FILE}. Cause: [405] during [GET] to [https://aae-3734-env.envalfresco.com/alfresco/api/-default-/public/alfresco/versions/1/nodes/] [NodesApiClient#getNode(String,List,String,List)]: [{\"error\":{\"errorKey\":\"framework.exception.UnsupportedResourceOperation\",\"statusCode\":405,\"briefSummary\":\"09070282 The operation is unsupported\",\"stackTrace\":\"For security reasons the stack trace is no longer displayed, but the property is kept for previous versions\",\"descriptionURL\":\"https://api-explorer.alfresco.com\"}}]";
@@ -73,9 +95,12 @@ public class QueryAdminProcessServiceTasksIT {
     private static final ParameterizedTypeReference<PagedModel<CloudServiceTask>> PAGED_TASKS_RESPONSE_TYPE = new ParameterizedTypeReference<PagedModel<CloudServiceTask>>() {
     };
 
-    private static final ParameterizedTypeReference<CloudServiceTask> SINGLE_TASK_RESPONSE_TYPE = new ParameterizedTypeReference<CloudServiceTask>() { };
+    private static final ParameterizedTypeReference<CloudServiceTask> SINGLE_TASK_RESPONSE_TYPE = new ParameterizedTypeReference<CloudServiceTask>() {
+    };
 
-    private static final ParameterizedTypeReference<CloudIntegrationContext> SINGLE_INT_CONTEXT_RESPONSE_TYPE = new ParameterizedTypeReference<CloudIntegrationContext>() { };
+    private static final ParameterizedTypeReference<CloudIntegrationContext> SINGLE_INT_CONTEXT_RESPONSE_TYPE = new ParameterizedTypeReference<CloudIntegrationContext>() {
+    };
+    public static final String SERVICE_TASKS_URL = "/admin/v1/service-tasks";
 
     @Autowired
     private IdentityTokenProducer identityTokenProducer;
@@ -106,6 +131,8 @@ public class QueryAdminProcessServiceTasksIT {
 
     private String processDefinitionId = UUID.randomUUID().toString();
 
+    private ProcessInstanceEventContainedBuilder processInstanceBuilder;
+
 
     private EventsAggregator eventsAggregator;
 
@@ -114,6 +141,8 @@ public class QueryAdminProcessServiceTasksIT {
         identityTokenProducer.setTestUser("hradmin");
 
         eventsAggregator = new EventsAggregator(producer);
+
+        processInstanceBuilder = new ProcessInstanceEventContainedBuilder(eventsAggregator);
 
         //given
         ProcessDefinitionImpl firstProcessDefinition = new ProcessDefinitionImpl();
@@ -141,10 +170,7 @@ public class QueryAdminProcessServiceTasksIT {
     @Test
     public void shouldGetProcessInstanceServiceTasks() throws InterruptedException {
         //given
-        ProcessInstanceImpl process = startSimpleProcessInstance();
-
-        //when
-        eventsAggregator.sendAll();
+        ProcessInstanceImpl process = sendEventsForStartSimpleProcessInstance();
 
         //then
         await().untilAsserted(() -> {
@@ -170,10 +196,7 @@ public class QueryAdminProcessServiceTasksIT {
     @Test
     public void shouldGetProcessInstanceServiceTasksByStatus() throws InterruptedException {
         //given
-        ProcessInstanceImpl process = startSimpleProcessInstance();
-
-        //when
-        eventsAggregator.sendAll();
+        ProcessInstanceImpl process = sendEventsForStartSimpleProcessInstance();
 
         //then
         await().untilAsserted(() -> {
@@ -192,8 +215,8 @@ public class QueryAdminProcessServiceTasksIT {
             assertThat(responseEntity.getStatusCode()).isEqualTo(HttpStatus.OK);
             assertThat(responseEntity.getBody()).isNotNull();
             assertThat(responseEntity.getBody().getContent()).hasSize(1)
-                                                             .extracting(CloudServiceTask::getStatus, CloudServiceTask::getActivityType)
-                                                             .contains(tuple(CloudBPMNActivity.BPMNActivityStatus.STARTED, SERVICE_TASK_TYPE));
+                .extracting(CloudServiceTask::getStatus, CloudServiceTask::getActivityType)
+                .contains(tuple(CloudBPMNActivity.BPMNActivityStatus.STARTED, SERVICE_TASK_TYPE));
         });
 
         // and given
@@ -202,9 +225,7 @@ public class QueryAdminProcessServiceTasksIT {
         taskActivity.setProcessInstanceId(process.getId());
         taskActivity.setExecutionId(UUID.randomUUID().toString());
 
-        eventsAggregator.addEvents(new CloudBPMNActivityCompletedEventImpl(taskActivity, processDefinitionId, process.getId()));
-
-        eventsAggregator.sendAll();
+        sendActivityCompletedEvent(taskActivity, process);
 
         await().untilAsserted(() -> {
             //when
@@ -217,18 +238,15 @@ public class QueryAdminProcessServiceTasksIT {
             assertThat(responseEntity.getStatusCode()).isEqualTo(HttpStatus.OK);
             assertThat(responseEntity.getBody()).isNotNull();
             assertThat(responseEntity.getBody().getContent()).hasSize(1)
-                                                             .extracting(CloudServiceTask::getStatus, CloudServiceTask::getActivityType)
-                                                             .contains(tuple(CloudBPMNActivity.BPMNActivityStatus.COMPLETED, SERVICE_TASK_TYPE));
+                .extracting(CloudServiceTask::getStatus, CloudServiceTask::getActivityType)
+                .contains(tuple(CloudBPMNActivity.BPMNActivityStatus.COMPLETED, SERVICE_TASK_TYPE));
         });
     }
 
     @Test
     public void shouldGetServiceTasks() throws InterruptedException {
         //given
-        ProcessInstanceImpl process = startSimpleProcessInstance();
-
-        //when
-        eventsAggregator.sendAll();
+        ProcessInstanceImpl process = sendEventsForStartSimpleProcessInstance();
 
         //then
         await().untilAsserted(() -> {
@@ -238,27 +256,21 @@ public class QueryAdminProcessServiceTasksIT {
 
         await().untilAsserted(() -> {
             //when
-            ResponseEntity<PagedModel<CloudServiceTask>> responseEntity = testRestTemplate.exchange("/admin/v1/service-tasks",
-                                                                                       HttpMethod.GET,
-                                                                                       identityTokenProducer.entityWithAuthorizationHeader(),
-                                                                                       PAGED_TASKS_RESPONSE_TYPE);
+            ResponseEntity<PagedModel<CloudServiceTask>> responseEntity = testRestTemplate.exchange(SERVICE_TASKS_URL,
+                HttpMethod.GET, identityTokenProducer.entityWithAuthorizationHeader(), PAGED_TASKS_RESPONSE_TYPE);
             //then
             assertThat(responseEntity.getStatusCode()).isEqualTo(HttpStatus.OK);
             assertThat(responseEntity.getBody()).isNotNull();
             assertThat(responseEntity.getBody().getContent()).hasSize(1)
-                                                             .extracting(CloudServiceTask::getActivityType)
-                                                             .contains(SERVICE_TASK_TYPE);
+                .extracting(CloudServiceTask::getActivityType)
+                .contains(SERVICE_TASK_TYPE);
         });
     }
 
     @Test
     public void shouldGetServiceTasksByStatus() throws InterruptedException {
         //given
-        ProcessInstanceImpl process = startSimpleProcessInstance();
-
-
-        //when
-        eventsAggregator.sendAll();
+        ProcessInstanceImpl process = sendEventsForStartSimpleProcessInstance();
 
         //then
         await().untilAsserted(() -> {
@@ -287,9 +299,7 @@ public class QueryAdminProcessServiceTasksIT {
         taskActivity.setProcessInstanceId(process.getId());
         taskActivity.setExecutionId(UUID.randomUUID().toString());
 
-        eventsAggregator.addEvents(new CloudBPMNActivityCompletedEventImpl(taskActivity, processDefinitionId, process.getId()));
-
-        eventsAggregator.sendAll();
+        sendActivityCompletedEvent(taskActivity, process);
 
         await().untilAsserted(() -> {
             //when
@@ -311,10 +321,7 @@ public class QueryAdminProcessServiceTasksIT {
     @Test
     public void shouldGetServiceTaskById() throws InterruptedException {
         //given
-        ProcessInstanceImpl process = startSimpleProcessInstance();
-
-        //when
-        eventsAggregator.sendAll();
+        ProcessInstanceImpl process = sendEventsForStartSimpleProcessInstance();
 
         //then
         await().untilAsserted(() -> {
@@ -324,10 +331,8 @@ public class QueryAdminProcessServiceTasksIT {
 
         await().untilAsserted(() -> {
 
-            ResponseEntity<PagedModel<CloudServiceTask>> serviceTasksResponse = testRestTemplate.exchange("/admin/v1/service-tasks",
-                                                                                                      HttpMethod.GET,
-                                                                                                      identityTokenProducer.entityWithAuthorizationHeader(),
-                                                                                                      PAGED_TASKS_RESPONSE_TYPE);
+            ResponseEntity<PagedModel<CloudServiceTask>> serviceTasksResponse = testRestTemplate.exchange(SERVICE_TASKS_URL,
+                HttpMethod.GET, identityTokenProducer.entityWithAuthorizationHeader(), PAGED_TASKS_RESPONSE_TYPE);
 
             assertThat(serviceTasksResponse.getBody().getContent()).isNotEmpty();
 
@@ -353,10 +358,7 @@ public class QueryAdminProcessServiceTasksIT {
     @Test
     public void shouldGetServiceTaskIntegrationContextErrorById() throws InterruptedException {
         //given
-        ProcessInstanceImpl process = startSimpleProcessInstance();
-
-        //when
-        eventsAggregator.sendAll();
+        ProcessInstanceImpl process = sendEventsForStartSimpleProcessInstance();
 
         //then
         await().untilAsserted(() -> {
@@ -364,10 +366,8 @@ public class QueryAdminProcessServiceTasksIT {
             assertThat(bpmnSequenceFlowRepository.findByProcessInstanceId(process.getId())).hasSize(1);
         });
 
-        ResponseEntity<PagedModel<CloudServiceTask>> serviceTasksResponse = testRestTemplate.exchange("/admin/v1/service-tasks",
-                                                                                                       HttpMethod.GET,
-                                                                                                       identityTokenProducer.entityWithAuthorizationHeader(),
-                                                                                                       PAGED_TASKS_RESPONSE_TYPE);
+        ResponseEntity<PagedModel<CloudServiceTask>> serviceTasksResponse = testRestTemplate.exchange(SERVICE_TASKS_URL,
+            HttpMethod.GET, identityTokenProducer.entityWithAuthorizationHeader(), PAGED_TASKS_RESPONSE_TYPE);
 
         assertThat(serviceTasksResponse.getBody().getContent()).isNotEmpty();
 
@@ -377,32 +377,17 @@ public class QueryAdminProcessServiceTasksIT {
                                                             .next();
 
         final String rootProcessInstanceId = UUID.randomUUID().toString();
-        IntegrationContextImpl integrationContext = new IntegrationContextImpl();
-        integrationContext.setProcessInstanceId(process.getId());
-        integrationContext.setRootProcessInstanceId(rootProcessInstanceId);
-        integrationContext.setExecutionId(serviceTask.getExecutionId());
-        integrationContext.setClientId(serviceTask.getElementId());
-        integrationContext.setClientType(serviceTask.getActivityType());
-        integrationContext.setClientName(serviceTask.getActivityName());
-        integrationContext.setProcessDefinitionId(process.getProcessDefinitionId());
-        integrationContext.addInBoundVariable("key", "value");
+        final IntegrationContext integrationContext = buildIntegrationContext(process,
+            rootProcessInstanceId, serviceTask);
 
-        eventsAggregator.addEvents(new CloudIntegrationRequestedEventImpl(integrationContext));
+        sendIntegrationRequestedEvent(integrationContext);
 
-        eventsAggregator.sendAll();
-
-         //when
+        //when
         await().untilAsserted(() -> {
 
-            ResponseEntity<CloudIntegrationContext> responseEntity = testRestTemplate.exchange("/admin/v1/service-tasks/{serviceTaskId}/integration-context",
-                                                                                               HttpMethod.GET,
-                                                                                               identityTokenProducer.entityWithAuthorizationHeader(),
-                                                                                               SINGLE_INT_CONTEXT_RESPONSE_TYPE,
-                                                                                               serviceTask.getId());
-            //then
-            assertThat(responseEntity.getStatusCode()).isEqualTo(HttpStatus.OK);
-            assertThat(responseEntity.getBody()).isNotNull();
-            assertThat(responseEntity.getBody())
+            CloudIntegrationContext cloudIntegrationContext = retrieveIntegrationContext(
+                serviceTask.getId());
+            assertThat(cloudIntegrationContext)
                 .extracting(
                     CloudIntegrationContext::getClientId,
                     CloudIntegrationContext::getClientType,
@@ -430,61 +415,29 @@ public class QueryAdminProcessServiceTasksIT {
 
         await().untilAsserted(() -> {
 
-            ResponseEntity<CloudIntegrationContext> responseEntity = testRestTemplate.exchange("/admin/v1/service-tasks/{serviceTaskId}/integration-context",
-                                                                                               HttpMethod.GET,
-                                                                                               identityTokenProducer.entityWithAuthorizationHeader(),
-                                                                                               SINGLE_INT_CONTEXT_RESPONSE_TYPE,
-                                                                                               serviceTask.getId());
-            //then
-            assertThat(responseEntity.getStatusCode()).isEqualTo(HttpStatus.OK);
-            assertThat(responseEntity.getBody()).isNotNull();
-            assertThat(responseEntity.getBody()).extracting(CloudIntegrationContext::getClientId,
-                                                            CloudIntegrationContext::getClientType,
-                                                            CloudIntegrationContext::getStatus,
-                                                            CloudIntegrationContext::getErrorCode,
-                                                            CloudIntegrationContext::getErrorMessage,
-                                                            CloudIntegrationContext::getErrorClassName)
-                                                .containsExactly(SERVICE_TASK_ELEMENT_ID,
-                                                                 SERVICE_TASK_TYPE,
-                                                                 IntegrationContextStatus.INTEGRATION_ERROR_RECEIVED,
-                                                                 error.getErrorCode(),
-                                                                 StringUtils.truncate(error.getMessage(),
-                                                                                      ERROR_MESSAGE_LENGTH),
-                                                                 error.getClass().getName());
+            CloudIntegrationContext cloudIntegrationContext = retrieveIntegrationContext(
+                serviceTask.getId());
+            assertThat(cloudIntegrationContext).extracting(CloudIntegrationContext::getClientId,
+                    CloudIntegrationContext::getClientType,
+                    CloudIntegrationContext::getStatus,
+                    CloudIntegrationContext::getErrorCode,
+                    CloudIntegrationContext::getErrorMessage,
+                    CloudIntegrationContext::getErrorClassName)
+                .containsExactly(SERVICE_TASK_ELEMENT_ID,
+                    SERVICE_TASK_TYPE,
+                    IntegrationContextStatus.INTEGRATION_ERROR_RECEIVED,
+                    error.getErrorCode(),
+                    StringUtils.truncate(error.getMessage(),
+                        ERROR_MESSAGE_LENGTH),
+                    error.getClass().getName());
 
-            assertThat(responseEntity.getBody().getStackTraceElements()).isNotEmpty();
+            assertThat(cloudIntegrationContext.getStackTraceElements()).isNotEmpty();
 
         });
     }
 
-    @Test
-    public void shouldGetServiceTaskIntegrationContextResultById() throws InterruptedException {
-        //given
-        ProcessInstanceImpl process = startSimpleProcessInstance();
-
-        //when
-        eventsAggregator.sendAll();
-
-        //then
-        await()
-            .untilAsserted(() -> {
-                assertThat(bpmnActivityRepository.findByProcessInstanceId(process.getId())).hasSize(2);
-                assertThat(bpmnSequenceFlowRepository.findByProcessInstanceId(process.getId())).hasSize(1);
-            });
-
-        ResponseEntity<PagedModel<CloudServiceTask>> serviceTasksResponse = testRestTemplate.exchange("/admin/v1/service-tasks",
-                                                                                                      HttpMethod.GET,
-                                                                                                      identityTokenProducer.entityWithAuthorizationHeader(),
-                                                                                                      PAGED_TASKS_RESPONSE_TYPE);
-
-        assertThat(serviceTasksResponse.getBody().getContent()).isNotEmpty();
-
-        CloudBPMNActivity serviceTask = serviceTasksResponse.getBody()
-                                                            .getContent()
-                                                            .iterator()
-                                                            .next();
-
-        final String rootProcessInstanceId = UUID.randomUUID().toString();
+    private IntegrationContextImpl buildIntegrationContext(ProcessInstance process,
+        String rootProcessInstanceId, CloudBPMNActivity serviceTask) {
         IntegrationContextImpl integrationContext = new IntegrationContextImpl();
         integrationContext.setProcessInstanceId(process.getId());
         integrationContext.setRootProcessInstanceId(rootProcessInstanceId);
@@ -494,59 +447,178 @@ public class QueryAdminProcessServiceTasksIT {
         integrationContext.setClientName(serviceTask.getActivityName());
         integrationContext.setProcessDefinitionId(process.getProcessDefinitionId());
         integrationContext.addInBoundVariable("key", "value");
+        return integrationContext;
+    }
 
-        eventsAggregator.addEvents(new CloudIntegrationRequestedEventImpl(integrationContext));
+    @Test
+    public void shouldGetServiceTaskIntegrationContextResultById() throws InterruptedException {
+        //given
+        ProcessInstanceImpl process = sendEventsForStartSimpleProcessInstance();
 
-        eventsAggregator.sendAll();
+        //then
+        CloudServiceTask serviceTask = waitForServiceTask();
+        final String rootProcessInstanceId = UUID.randomUUID().toString();
+        IntegrationContext integrationContext = buildIntegrationContext(process,
+            rootProcessInstanceId,
+            serviceTask);
+
+        sendIntegrationRequestedEvent(integrationContext);
 
         //when
         await().untilAsserted(() -> {
 
-                ResponseEntity<CloudIntegrationContext> responseEntity = testRestTemplate.exchange("/admin/v1/service-tasks/{serviceTaskId}/integration-context",
-                                                                                                   HttpMethod.GET,
-                                                                                                   identityTokenProducer.entityWithAuthorizationHeader(),
-                                                                                                   SINGLE_INT_CONTEXT_RESPONSE_TYPE,
-                                                                                                   serviceTask.getId());
-                //then
-                assertThat(responseEntity.getStatusCode()).isEqualTo(HttpStatus.OK);
-                assertThat(responseEntity.getBody()).isNotNull();
-                assertThat(responseEntity.getBody())
-                    .extracting(
-                        CloudIntegrationContext::getClientId,
-                        CloudIntegrationContext::getClientType,
-                        CloudIntegrationContext::getRootProcessInstanceId,
-                        CloudIntegrationContext::getStatus)
-                    .containsExactly(
-                        SERVICE_TASK_ELEMENT_ID,
-                        SERVICE_TASK_TYPE,
-                        rootProcessInstanceId,
-                        IntegrationContextStatus.INTEGRATION_REQUESTED);
-            });
+            CloudIntegrationContext cloudIntegrationContext = retrieveIntegrationContext(
+                serviceTask.getId());
+            assertThat(cloudIntegrationContext)
+                .extracting(
+                    CloudIntegrationContext::getClientId,
+                    CloudIntegrationContext::getClientType,
+                    CloudIntegrationContext::getRootProcessInstanceId,
+                    CloudIntegrationContext::getStatus)
+                .containsExactly(
+                    SERVICE_TASK_ELEMENT_ID,
+                    SERVICE_TASK_TYPE,
+                    rootProcessInstanceId,
+                    IntegrationContextStatus.INTEGRATION_REQUESTED);
+        });
 
         // and given
-        eventsAggregator.addEvents(new CloudIntegrationResultReceivedEventImpl(integrationContext));
-
-        eventsAggregator.sendAll();
+        sendIntegrationResultReceivedEvent(integrationContext);
 
         await()
             .untilAsserted(() -> {
 
-                ResponseEntity<CloudIntegrationContext> responseEntity = testRestTemplate.exchange("/admin/v1/service-tasks/{serviceTaskId}/integration-context",
-                                                                                                   HttpMethod.GET,
-                                                                                                   identityTokenProducer.entityWithAuthorizationHeader(),
-                                                                                                   SINGLE_INT_CONTEXT_RESPONSE_TYPE,
-                                                                                                   serviceTask.getId());
-                //then
-                assertThat(responseEntity.getStatusCode()).isEqualTo(HttpStatus.OK);
-                assertThat(responseEntity.getBody()).isNotNull();
-                assertThat(responseEntity.getBody()).extracting(CloudIntegrationContext::getClientId,
-                                                                CloudIntegrationContext::getClientType,
-                                                                CloudIntegrationContext::getStatus)
-                                                    .containsExactly(SERVICE_TASK_ELEMENT_ID,
-                                                                     SERVICE_TASK_TYPE,
-                                                                     IntegrationContextStatus.INTEGRATION_RESULT_RECEIVED);
+                CloudIntegrationContext cloudIntegrationContext = retrieveIntegrationContext(serviceTask.getId());
+                assertThat(cloudIntegrationContext).extracting(
+                        CloudIntegrationContext::getClientId,
+                        CloudIntegrationContext::getClientType,
+                        CloudIntegrationContext::getStatus)
+                    .containsExactly(SERVICE_TASK_ELEMENT_ID,
+                        SERVICE_TASK_TYPE,
+                        IntegrationContextStatus.INTEGRATION_RESULT_RECEIVED);
             });
 
+    }
+
+    private CloudIntegrationContext retrieveIntegrationContext(String serviceTaskId) {
+        ResponseEntity<CloudIntegrationContext> responseEntity = testRestTemplate.exchange(
+            "/admin/v1/service-tasks/{serviceTaskId}/integration-context",
+            HttpMethod.GET,
+            identityTokenProducer.entityWithAuthorizationHeader(),
+            SINGLE_INT_CONTEXT_RESPONSE_TYPE,
+            serviceTaskId);
+        assertThat(responseEntity.getStatusCode()).isEqualTo(HttpStatus.OK);
+        assertThat(responseEntity.getBody()).isNotNull();
+        return responseEntity.getBody();
+    }
+
+    private void sendIntegrationResultReceivedEvent(IntegrationContext integrationContext) {
+        eventsAggregator.addEvents(new CloudIntegrationResultReceivedEventImpl(integrationContext));
+
+        eventsAggregator.sendAll();
+    }
+
+    private void sendIntegrationRequestedEvent(IntegrationContext integrationContext) {
+        eventsAggregator.addEvents(new CloudIntegrationRequestedEventImpl(integrationContext));
+
+        eventsAggregator.sendAll();
+    }
+
+    private CloudServiceTask waitForServiceTask(BPMNActivityStatus status) {
+        await()
+            .untilAsserted(() -> {
+                final PagedModel<CloudServiceTask> page = testRestTemplate.exchange(
+                    SERVICE_TASKS_URL, HttpMethod.GET,
+                    identityTokenProducer.entityWithAuthorizationHeader(),
+                    PAGED_TASKS_RESPONSE_TYPE).getBody();
+                assertThat(page)
+                    .isNotEmpty();
+                final CloudServiceTask serviceTask = page.getContent().iterator().next();
+                assertThat(serviceTask.getStatus()).isEqualTo(status);
+            });
+        return retrieveServiceTask();
+    }
+
+    private CloudServiceTask waitForServiceTask() {
+        return waitForServiceTask(BPMNActivityStatus.STARTED);
+    }
+
+    private CloudServiceTask retrieveServiceTask() {
+        ResponseEntity<PagedModel<CloudServiceTask>> serviceTasksResponse = testRestTemplate.exchange(
+            SERVICE_TASKS_URL,
+            HttpMethod.GET,
+            identityTokenProducer.entityWithAuthorizationHeader(),
+            PAGED_TASKS_RESPONSE_TYPE);
+
+        assertThat(serviceTasksResponse.getBody()).isNotEmpty();
+
+        return serviceTasksResponse.getBody()
+            .getContent()
+            .iterator()
+            .next();
+    }
+
+
+    @Test
+    public void should_supportLoopInvolvingServiceTasks() {
+        //given
+        ProcessInstanceImpl process = sendEventsForStartSimpleProcessInstance();
+        CloudServiceTask serviceTaskIt1 = waitForServiceTask();
+
+        final String rootProcessInstanceId = UUID.randomUUID().toString();
+        IntegrationContextImpl integrationContext = buildIntegrationContext(process,
+            rootProcessInstanceId,
+            serviceTaskIt1);
+
+        sendIntegrationRequestedEvent(integrationContext);
+        sendIntegrationResultReceivedEvent(integrationContext);
+        sendActivityCompletedEvent(serviceTaskIt1, process);
+
+        waitForServiceTask(BPMNActivityStatus.COMPLETED);
+        waitForIntegrationContext(serviceTaskIt1, IntegrationContextStatus.INTEGRATION_RESULT_RECEIVED);
+
+        //when the process loop back and reaches the task a second time
+        final BPMNActivityImpl bpmnActivity = buildServiceTask(serviceTaskIt1.getExecutionId(),
+            process);
+        sendActivitiStartedEvent(process, bpmnActivity);
+
+        final CloudServiceTask serviceTask = waitForServiceTask(BPMNActivityStatus.STARTED);
+        IntegrationContextImpl integrationContextIt2 = buildIntegrationContext(process, rootProcessInstanceId, serviceTask);
+        sendIntegrationRequestedEvent(integrationContextIt2);
+
+        //then
+        waitForIntegrationContext(serviceTaskIt1, IntegrationContextStatus.INTEGRATION_REQUESTED);
+    }
+
+    private void sendActivitiStartedEvent(ProcessInstanceImpl process, BPMNActivityImpl bpmnActivity) {
+        eventsAggregator.addEvents(
+            new CloudBPMNActivityStartedEventImpl(bpmnActivity, process.getProcessDefinitionId(),
+                process.getId()));
+
+        eventsAggregator.sendAll();
+    }
+
+    private void sendActivityCompletedEvent(BPMNActivity bpmnActivity, ProcessInstance processInstance) {
+        final CloudBPMNActivityCompletedEventImpl activityCompletedEvent = new CloudBPMNActivityCompletedEventImpl(
+            bpmnActivity, processInstance.getProcessDefinitionId(),
+            processInstance.getId());
+        eventsAggregator.addEvents(activityCompletedEvent);
+        eventsAggregator.sendAll();
+    }
+
+    private void waitForIntegrationContext(CloudServiceTask serviceTask, IntegrationContextStatus status) {
+        await()
+            .untilAsserted(() -> {
+                CloudIntegrationContext cloudIntegrationContext = retrieveIntegrationContext(
+                    serviceTask.getId());
+                assertThat(cloudIntegrationContext.getStatus()).isEqualTo(status);
+            });
+    }
+
+    private ProcessInstanceImpl sendEventsForStartSimpleProcessInstance() {
+        ProcessInstanceImpl process = startSimpleProcessInstance();
+        eventsAggregator.sendAll();
+        return process;
     }
 
 
@@ -555,10 +627,7 @@ public class QueryAdminProcessServiceTasksIT {
         //given
         identityTokenProducer.setTestUser("hruser");
 
-        ProcessInstanceImpl process = startSimpleProcessInstance();
-
-        //when
-        eventsAggregator.sendAll();
+        ProcessInstanceImpl process = sendEventsForStartSimpleProcessInstance();
 
         //then
         await().untilAsserted(() -> {
@@ -577,7 +646,8 @@ public class QueryAdminProcessServiceTasksIT {
         });
     }
 
-    protected ProcessInstanceImpl startSimpleProcessInstance() {
+    private ProcessInstanceImpl startSimpleProcessInstance() {
+        String executionId = UUID.randomUUID().toString();
         ProcessInstanceImpl process = new ProcessInstanceImpl();
         process.setId(UUID.randomUUID().toString());
         process.setName("process");
@@ -588,26 +658,32 @@ public class QueryAdminProcessServiceTasksIT {
         BPMNActivityImpl startActivity = new BPMNActivityImpl("startEvent1", "", "startEvent");
         startActivity.setProcessDefinitionId(process.getProcessDefinitionId());
         startActivity.setProcessInstanceId(process.getId());
-        startActivity.setExecutionId(UUID.randomUUID().toString());
+        startActivity.setExecutionId(executionId);
 
         BPMNSequenceFlowImpl sequenceFlow = new BPMNSequenceFlowImpl("sid-68945AF1-396F-4B8A-B836-FC318F62313F", "startEvent1", SERVICE_TASK_ELEMENT_ID);
         sequenceFlow.setProcessDefinitionId(process.getProcessDefinitionId());
         sequenceFlow.setProcessInstanceId(process.getId());
 
-        BPMNActivityImpl taskActivity = new BPMNActivityImpl(SERVICE_TASK_ELEMENT_ID, "Service Task", SERVICE_TASK_TYPE);
-        taskActivity.setProcessDefinitionId(process.getProcessDefinitionId());
-        taskActivity.setProcessInstanceId(process.getId());
-        taskActivity.setExecutionId(UUID.randomUUID().toString());
+        BPMNActivityImpl activity = buildServiceTask(executionId, process);
 
         eventsAggregator.addEvents(new CloudProcessCreatedEventImpl(process),
-                                   new CloudProcessStartedEventImpl(process, null, null),
-                                   new CloudBPMNActivityStartedEventImpl(startActivity, processDefinitionId, process.getId()),
-                                   new CloudBPMNActivityCompletedEventImpl(startActivity, processDefinitionId, process.getId()),
-                                   new CloudSequenceFlowTakenEventImpl(sequenceFlow),
-                                   new CloudBPMNActivityStartedEventImpl(taskActivity, processDefinitionId, process.getId())
+            new CloudProcessStartedEventImpl(process, null, null),
+            new CloudBPMNActivityStartedEventImpl(startActivity, processDefinitionId, process.getId()),
+            new CloudBPMNActivityCompletedEventImpl(startActivity, processDefinitionId, process.getId()),
+            new CloudSequenceFlowTakenEventImpl(sequenceFlow),
+            new CloudBPMNActivityStartedEventImpl(activity, processDefinitionId, process.getId())
         );
 
         return process;
 
+    }
+
+    private BPMNActivityImpl buildServiceTask(String executionId, ProcessInstanceImpl process) {
+        BPMNActivityImpl activity = new BPMNActivityImpl(SERVICE_TASK_ELEMENT_ID, "Service Task",
+            SERVICE_TASK_TYPE);
+        activity.setProcessDefinitionId(process.getProcessDefinitionId());
+        activity.setProcessInstanceId(process.getId());
+        activity.setExecutionId(executionId);
+        return activity;
     }
 }

--- a/activiti-cloud-query-service/activiti-cloud-starter-query/src/test/resources/application.properties
+++ b/activiti-cloud-query-service/activiti-cloud-starter-query/src/test/resources/application.properties
@@ -28,6 +28,6 @@ authorizations.security-constraints[0].securityCollections[0].patterns[0]=/v1/*
 authorizations.security-constraints[1].authRoles[0]=ACTIVITI_ADMIN
 authorizations.security-constraints[1].securityCollections[0].patterns[0]=/admin/*
 
-spring.jpa.properties.hibernate.generate_statistics=true
+spring.jpa.properties.hibernate.generate_statistics=false
 spring.jpa.properties.hibernate.format_sql=true
-spring.jpa.show-sql=true
+spring.jpa.show-sql=false

